### PR TITLE
1802 provide a way to remove clusters from the list

### PR DESCRIPTION
**Pull request by msivasubramaniaan**
Signed-off-by: msivasubramaniaan [msivasub@redhat.com](mailto:msivasub@redhat.com)

This will fix https://github.com/redhat-developer/vscode-openshift-tools/issues/1802

**Post Fix:**
delete-cluster-demo

### CODE REVIEW

**Comment by mohitsuman** [comment]
@msivasubramaniaan what exactly changed between commit [ad6dda7](https://github.com/redhat-developer/vscode-openshift-tools/pull/2678/commits/ad6dda752ae809172f39b3313b6dafd474eab6f1) & [9945c2a](https://github.com/redhat-developer/vscode-openshift-tools/pull/2678/commits/9945c2aa0739f17ff9dcfc41e2f818f730c0fbdc) .

**Comment by msivasubramaniaan** [comment]
> @msivasubramaniaan what exactly changed between commit [ad6dda7](https://github.com/redhat-developer/vscode-openshift-tools/pull/2678/commits/ad6dda752ae809172f39b3313b6dafd474eab6f1) & [9945c2a](https://github.com/redhat-developer/vscode-openshift-tools/pull/2678/commits/9945c2aa0739f17ff9dcfc41e2f818f730c0fbdc) .

wrongly committed the .editorConfig file, so reverted and add the particular files

File: src/openshift/cluster.ts [hunk 1]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+            const addBtn = new quickBtn(new ThemeIcon('plus'), 'Add');
+            const switchBtn = new quickBtn(new ThemeIcon('arrow-swap'), 'Switch');
+            const deleteBtn = new quickBtn(new ThemeIcon('trash'), 'Delete');
+            const createUrl: QuickPickItem = { label: 'Provide new URL...', buttons: [addBtn] };
+            const clusterItems = k8sConfig.getServers();
+            const quickPick = window.createQuickPick();
+            const contextNames: QuickPickItem[] = clusterItems.map((ctx) => ({ label: `${ctx.label}`, buttons: [switchBtn, deleteBtn] }));
+            quickPick.items = [createUrl, ...contextNames];
+            quickPick.onDidTriggerItemButton(async (event) => {
+                if (event.button instanceof quickBtn) {
```
Comment (dgolovin) [inline src/openshift/cluster.ts:143]: This instanceof check is not required.

File: src/openshift/cluster.ts [hunk 2]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+    private static async showQuikPick(clusterURl: string): Promise<string> {
+        return await new Promise<string | null>((resolve, reject) => {
+            const k8sConfig = new KubeConfigUtils();
+            const addBtn = new quickBtn(new ThemeIcon('plus'), 'Add');
+            const switchBtn = new quickBtn(new ThemeIcon('arrow-swap'), 'Switch');
+            const deleteBtn = new quickBtn(new ThemeIcon('trash'), 'Delete');
+            const createUrl: QuickPickItem = { label: 'Provide new URL...', buttons: [addBtn] };
+            const clusterItems = k8sConfig.getServers();
+            const quickPick = window.createQuickPick();
+            const contextNames: QuickPickItem[] = clusterItems.map((ctx) => ({ label: `${ctx.label}`, buttons: [switchBtn, deleteBtn] }));
```
Comment (dgolovin) [inline src/openshift/cluster.ts:140]: No need to have **switchBtn**, just recognize selection of cluster in **onDidAccept** listener.

File: src/openshift/cluster.ts [hunk 3]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+        return await Cluster.showQuikPick(clusterURl);
+    }
+
+    private static async showQuikPick(clusterURl: string): Promise<string> {
+        return await new Promise<string | null>((resolve, reject) => {
+            const k8sConfig = new KubeConfigUtils();
+            const addBtn = new quickBtn(new ThemeIcon('plus'), 'Add');
+            const switchBtn = new quickBtn(new ThemeIcon('arrow-swap'), 'Switch');
+            const deleteBtn = new quickBtn(new ThemeIcon('trash'), 'Delete');
+            const createUrl: QuickPickItem = { label: 'Provide new URL...', buttons: [addBtn] };
```
Comment (dgolovin) [inline src/openshift/cluster.ts:137]: **addBtn** is not needed, just recognize Add quickpick in **onDidAccept** listener.

File: src/openshift/cluster.ts [hunk 4]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+            const clusterItems = k8sConfig.getServers();
+            const quickPick = window.createQuickPick();
+            const contextNames: QuickPickItem[] = clusterItems.map((ctx) => ({ label: `${ctx.label}`, buttons: [switchBtn, deleteBtn] }));
+            quickPick.items = [createUrl, ...contextNames];
+            quickPick.onDidTriggerItemButton(async (event) => {
+                if (event.button instanceof quickBtn) {
+                    if (event.button.iconPath.id === 'plus') {
```
Comment (dgolovin) [inline src/openshift/cluster.ts:144]: Recognize **Provide New URL...** item in **onDidAccept** listener and move code from this if there.

File: src/openshift/cluster.ts [hunk 5]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+                        resolve(window.showInputBox({
+                            value: clusterURl,
+                            ignoreFocusOut: true,
+                            prompt: 'Provide new Cluster URL to connect',
+                            validateInput: (value: string) => Cluster.validateUrl('Invalid URL provided', value)
+                        }));
+                    } else if (event.button.iconPath.id === 'arrow-swap') {
```
Comment (dgolovin) [inline src/openshift/cluster.ts:151]: move this code in **onDidAccept** listener

File: src/openshift/cluster.ts [hunk 6]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+                        resolve(event.item.label);
+                    } else if (event.button.iconPath.id === 'trash') {
```
Comment (dgolovin) [inline src/openshift/cluster.ts:153]: Use direct comparison **event.button === deleteBtn**.

File: src/openshift/cluster.ts [hunk 7]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+                        await window.showInformationMessage('Are you sure want to delete the Cluster?', 'Yes', 'No')
+                            .then(async answer => {
+                                if (answer === 'Yes') {
+                                    const cluster = k8sConfig.getClusters().filter((kubeConfigCluster) => kubeConfigCluster.server === event.item.label).pop();
+                                    try {
+                                        await k8sConfig.deleteCluster(cluster);
+                                        resolve('');
+                                    } catch (err) {
+                                        reject(null);
```
Comment (dgolovin) [inline src/openshift/cluster.ts:162]: Show error message from the **err** exception.

File: src/openshift/cluster.ts [hunk 8]
```diff
@@ -120,19 +124,50 @@ export class Cluster extends OpenShiftItem {
...
+                        }));
+                    } else if (event.button.iconPath.id === 'arrow-swap') {
+                        resolve(event.item.label);
+                    } else if (event.button.iconPath.id === 'trash') {
+                        await window.showInformationMessage('Are you sure want to delete the Cluster?', 'Yes', 'No')
+                            .then(async answer => {
+                                if (answer === 'Yes') {
+                                    const cluster = k8sConfig.getClusters().filter((kubeConfigCluster) => kubeConfigCluster.server === event.item.label).pop();
+                                    try {
+                                        await k8sConfig.deleteCluster(cluster);
```
Comment (dgolovin) [inline src/openshift/cluster.ts:159]: use `CliChannel.getInstance().execTool(Command.deleteCluster(cluster.name)`

File: src/util/kubeUtils.ts [hunk 9]
```diff
@@ -8,6 +8,8 @@ import * as path from 'path';
 import { QuickPickItem } from 'vscode';
 import { KubeConfig, findHomeDir, loadYaml } from '@kubernetes/client-node';
 import { User, Cluster } from '@kubernetes/client-node/dist/config_types';
+import { OdoImpl } from '../odo';
```
Comment (dgolovin) [inline src/util/kubeUtils.ts:11]: No changes required in this file if **CliChannel** class is used (see above).

**Review by dgolovin — CHANGES_REQUESTED** [review]
There should be no item buttons for **Provide new URL...** item and switch item button for every server url. You ser should be selecting items from quick pick list as usual. There **onDidAccept** listener for that.

**Review by dgolovin — CHANGES_REQUESTED** [review]
Couple more changes and it should be good to go.

**Review by dgolovin — CHANGES_REQUESTED** [review]
`Houston we have a problem!`

After deleting the CRC cluster with

```
"/home/denis/p/vscode-openshift-tools/out/tools/linux/oc" config delete-cluster api-crc-testing:6443
deleted cluster api-crc-testing:6443 from /home/denis/.kube/config
```

two contexts left in config file:
* crc-developer
* crc-admin

both can be selected in Switch context command and that leads to Application Explorer view showing welcome page with buttons. This is not correct because user is not logged out of cluster or cluster is not accessible. It is because the config is broken and any command that need config fails with error like this:

```
"/home/denis/p/vscode-openshift-tools/out/tools/linux/oc" get project -o=json

error: Missing or incomplete configuration info.  Please point to an existing, complete config file:

  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config

To view or setup config directly use the 'config' command.
```

That leads to more problems after CRC restart:

```
INFO Operators are stable (3/3)...                
INFO Adding crc-admin and crc-developer contexts to kubeconfig... 
ERRO Cannot update kubeconfig: Head "https://oauth-openshift.apps-crc.testing": dial tcp 192.168.130.11:443: connect: connection refused 
Started the OpenShift cluster.
```

Which lead to CRC not being able to add CRC related objects to config because it is broken.

Considering all this we should ether use k8s javascript client to traverse the config structure and firs remove all related contexts, then all related users, finally cluster itself, or execute several commands 'oc  config delete-context/delete-user/delete-cluster'.

I am leaning to use oc command so we don't have to deal with configuration loaded from several config files configured in KUBECONFIG evn variable.

**Comment by msivasubramaniaan** [comment]
> `Houston we have a problem!`
> 
> After deleting the CRC cluster with
> 
> ```
> "/home/denis/p/vscode-openshift-tools/out/tools/linux/oc" config delete-cluster api-crc-testing:6443
> deleted cluster api-crc-testing:6443 from /home/denis/.kube/config
> ```
> 
> two contexts left in config file:
> 
> * crc-developer
> * crc-admin
> 
> both can be selected in Switch context command and that leads to Application Explorer view showing welcome page with buttons. This is not correct because user is not logged out of cluster or cluster is not accessible. It is because the config is broken and any command that need config fails with error like this:
> 
> ```
> "/home/denis/p/vscode-openshift-tools/out/tools/linux/oc" get project -o=json
> 
> error: Missing or incomplete configuration info.  Please point to an existing, complete config file:
> 
>   1. Via the command-line flag --kubeconfig
>   2. Via the KUBECONFIG environment variable
>   3. In your home directory as ~/.kube/config
> 
> To view or setup config directly use the 'config' command.
> ```
> 
> That leads to more problems after CRC restart:
> 
> ```
> INFO Operators are stable (3/3)...                
> INFO Adding crc-admin and crc-developer contexts to kubeconfig... 
> ERRO Cannot update kubeconfig: Head "https://oauth-openshift.apps-crc.testing": dial tcp 192.168.130.11:443: connect: connection refused 
> Started the OpenShift cluster.
> ```
> 
> Which lead to CRC not being able to add CRC related objects to config because it is broken.
> 
> Considering all this we should ether use k8s javascript client to traverse the config structure and firs remove all related contexts, then all related users, finally cluster itself, or execute several commands 'oc config delete-context/delete-user/delete-cluster'.
> 
> I am leaning to use oc command so we don't have to deal with configuration loaded from several config files configured in KUBECONFIG evn variable.

@dgolovin as per your comment, currently I am deleting the context, user and finally the cluster

**Review by dgolovin — APPROVED** [review]
Updated code and created commit per change with description what was done.

**Pull request merged**